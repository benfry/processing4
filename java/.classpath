<?xml version="1.0" encoding="UTF-8"?>
<classpath>
	<classpathentry kind="src" path="generated"/>
	<classpathentry kind="src" path="src"/>
	<classpathentry combineaccessrules="false" kind="src" path="/processing4-app"/>
	<classpathentry combineaccessrules="false" kind="src" path="/processing4-core"/>
	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.launching.macosx.MacOSXType/adoptopenjdk-11">
		<attributes>
			<attribute name="module" value="true"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="lib" path="mode/antlr-4.7.2-complete.jar"/>
	<classpathentry kind="lib" path="mode/build-flexmark-all.jar"/>
	<classpathentry kind="lib" path="mode/classpath-explorer-1.0.jar"/>
	<classpathentry kind="lib" path="mode/com.ibm.icu.jar"/>
	<classpathentry kind="lib" path="mode/jdtCompilerAdapter.jar"/>
	<classpathentry kind="lib" path="mode/jsoup-1.7.1.jar"/>
	<classpathentry kind="lib" path="mode/org.eclipse.core.contenttype.jar"/>
	<classpathentry kind="lib" path="mode/org.eclipse.core.jobs.jar"/>
	<classpathentry kind="lib" path="mode/org.eclipse.core.resources.jar"/>
	<classpathentry kind="lib" path="mode/org.eclipse.core.runtime.jar"/>
	<classpathentry kind="lib" path="mode/org.eclipse.equinox.common.jar"/>
	<classpathentry kind="lib" path="mode/org.eclipse.equinox.preferences.jar"/>
	<classpathentry kind="lib" path="mode/org.eclipse.jdt.core.jar"/>
	<classpathentry kind="lib" path="mode/org.eclipse.osgi.jar"/>
	<classpathentry kind="lib" path="mode/org.eclipse.text.jar"/>
	<classpathentry kind="lib" path="mode/org.netbeans.swing.outline.jar"/>
	<classpathentry kind="lib" path="mode/ST-4.3.4.jar"/>
	<classpathentry kind="lib" path="test/lib/junit-4.13.2.jar"/>
	<classpathentry kind="lib" path="test/lib/hamcrest-core-1.3.jar"/>
	<classpathentry kind="output" path="bin"/>
</classpath>
