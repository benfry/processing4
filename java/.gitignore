reference.zip
bin
bin-test
generated
mode/JavaMode.jar
