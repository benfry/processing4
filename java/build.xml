<?xml version="1.0"?>
<project name="Java Mode" default="build">

  <property name="generated"
            value="${basedir}/generated/processing/mode/java/preproc" />
  <mkdir dir="${generated}" />

  <property name="grammars"
            value="${basedir}/src/processing/mode/java/preproc" />

  <property name="antlr_jar"
            value="${basedir}/mode/antlr-4.7.2-complete.jar" />

  <property name="mode_jar"
            value="${basedir}/mode/JavaMode.jar" />

  <classloader taskname="antlr">
    <classpath path="${antlr_jar}" />
  </classloader>

  <target name="clean" description="Clean the build directories">
    <delete dir="bin" />
    <delete dir="bin-test" />
    <delete file="${mode_jar}" />
    <delete>
      <fileset dir="${generated}">
        <include name="*.java" />
        <include name="*.tokens" />
        <include name="*.txt" />
        <include name="*.g" />
        <include name="*.smap" />
        <include name="*.properties" />
      </fileset>
    </delete>
  </target>

  <target name="preproc" description="Compile ANTLR 4 grammar">
    <java classname="org.antlr.v4.Tool"
        failonerror="true">
      <classpath path="${antlr_jar}" />
      <arg value="-o" />
      <arg value="${generated}" />
      <arg value="-package" />
      <arg value="processing.mode.java.preproc" />
      <arg value="${grammars}/Processing.g4" />
    </java>
  </target>

  <!-- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -->

  <property name="lsp4j.version" value="0.19.0" />

  <fileset id="lsp4j.files" dir="mode">
    <include name="org.eclipse.lsp4j.jar" />
    <include name="org.eclipse.lsp4j.jsonrpc.jar" />
    <include name="gson.jar" />
  </fileset>

  <condition property="lsp4j.present">
    <resourcecount refid="lsp4j.files" when="eq" count="3" />
  </condition>

  <!-- ok to ignore failed downloads if we at least have a version that's local -->
  <condition property="lsp4j.ignorable" value="false" else="true">
    <isset property="lsp4j.present" />
  </condition>

  <target name="download-lsp4j" unless="lsp4j.present">
    <!-- TODO implement a fallback URL that points to a location
         on download.processing.org so it's available forever. -->
    <property name="lsp4j.prefix"
              value="https://repo1.maven.org/maven2/org/eclipse/lsp4j" />

    <get src="${lsp4j.prefix}/org.eclipse.lsp4j/${lsp4j.version}/org.eclipse.lsp4j-${lsp4j.version}.jar"
         dest="mode/org.eclipse.lsp4j.jar"
         ignoreerrors="${lsp4j.ignorable}"
         usetimestamp="true" />

    <get src="${lsp4j.prefix}/org.eclipse.lsp4j.jsonrpc/${lsp4j.version}/org.eclipse.lsp4j.jsonrpc-${lsp4j.version}.jar"
         dest="mode/org.eclipse.lsp4j.jsonrpc.jar"
         ignoreerrors="${lsp4j.ignorable}"
         usetimestamp="true" />

    <get src="https://repo1.maven.org/maven2/com/google/code/gson/gson/2.9.1/gson-2.9.1.jar"
         dest="mode/gson.jar"
         ignoreerrors="${lsp4j.ignorable}"
         usetimestamp="true" />
  </target>

  <!-- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -->

  <path id="classpath.base">
    <pathelement location="../core/library/core.jar" />
    <pathelement location="../app/pde.jar" />
    <pathelement location="../app/lib/ant.jar" />
    <pathelement location="../app/lib/ant-launcher.jar" />
    <pathelement location="../app/lib/apple.jar" />
    <pathelement location="../app/lib/jna.jar" />
    <pathelement location="../app/lib/jna-platform.jar" />
    <pathelement location="../app/lib/build-flexmark-all.jar" />
    <pathelement location="../app/lib/ST-4.3.4.jar" />


    <pathelement location="mode/antlr-4.7.2-complete.jar" />
    <pathelement location="mode/classpath-explorer-1.0.jar" />
    <pathelement location="mode/jsoup-1.7.1.jar" />
    <pathelement location="mode/org.netbeans.swing.outline.jar" />
    <pathelement location="mode/com.ibm.icu.jar" />

    <pathelement location="mode/org.eclipse.core.contenttype.jar" />
    <pathelement location="mode/org.eclipse.core.jobs.jar" />
    <pathelement location="mode/org.eclipse.core.resources.jar" />
    <pathelement location="mode/org.eclipse.core.runtime.jar" />
    <pathelement location="mode/org.eclipse.equinox.common.jar" />
    <pathelement location="mode/org.eclipse.equinox.preferences.jar" />
    <pathelement location="mode/org.eclipse.jdt.compiler.apt.jar" />
    <pathelement location="mode/org.eclipse.jdt.core.jar" />
    <pathelement location="mode/org.eclipse.lsp4j.jar" />
    <pathelement location="mode/org.eclipse.lsp4j.jsonrpc.jar" />
    <pathelement location="mode/org.eclipse.osgi.jar" />
    <pathelement location="mode/org.eclipse.text.jar" />

  </path>

  <path id="classpath.test">
    <pathelement location="../core/library-test/junit-4.13.2.jar" />
    <pathelement location="../core/library-test/hamcrest-core-1.3.jar" />
    <pathelement location="../core/library-test/byte-buddy-1.11.19.jar" />
    <pathelement location="../core/library-test/objenesis-3.2.jar" />
    <pathelement location="../core/library-test/mockito-core-4.0.0.jar" />
    <pathelement location = "bin-test" />
    <path refid="classpath.base" />
  </path>

  <macrodef name="compilecommon">
    <attribute name="destdir"/>
    <attribute name="srcdir"/>
    <attribute name="classpath"/>
    <sequential>
      <condition property="core-built">
        <available file="../core/library/core.jar" />
      </condition>
      <fail unless="core-built" message="Please first build the core library and make sure it is located at ../core/library/core.jar" />

      <condition property="app-built">
        <available file="../app/pde.jar" />
      </condition>
      <fail unless="app-built" message="Please build app first and make sure it is located at ../app/pde.jar" />

      <mkdir dir="@{destdir}" />

      <!-- in some cases, pde.jar was not getting built
         https://github.com/processing/processing/issues/1792 -->
      <delete file="${mode_jar}" />

      <!-- env used to set classpath below -->
      <property environment="env" />

      <javac source="17"
             target="17"
             destdir="@{destdir}"
             excludes="**/tools/format/**"
             encoding="UTF-8"
             includeAntRuntime="false"
             debug="on"
             nowarn="true"
             compiler="modern">
        <src path="@{srcdir}" />
        <src path="generated" />
        <classpath refid="@{classpath}"/>
      </javac>
    </sequential>
  </macrodef>

  <target name="test-compile" depends="preproc">
    <compilecommon srcdir="src; test/processing" destdir="bin-test" classpath="classpath.test" />
  </target>

  <target name="test" depends="build, test-compile">
    <junit haltonfailure="true">
      <classpath refid="classpath.test" />
      <formatter type="brief" usefile="false" />
      <batchtest>
        <fileset dir="test">
          <include name="**/*Test.java" />
          <include name="**/*Tests.java" />
        </fileset>
      </batchtest>
    </junit>
  </target>

  <target name="compile" description="Compile sources" depends="preproc,download-lsp4j">
    <compilecommon srcdir="src" destdir="bin" classpath="classpath.base" />
  </target>

  <target name="build" depends="compile" description="Build PDE">
    <jar basedir="bin" destfile="${mode_jar}" />
  </target>

</project>
