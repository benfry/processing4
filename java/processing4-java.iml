<?xml version="1.0" encoding="UTF-8"?>
<module type="JAVA_MODULE" version="4">
  <component name="EclipseModuleManager">
    <libelement value="jar://$MODULE_DIR$/mode/antlr-4.7.2-complete.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/classpath-explorer-1.0.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/com.ibm.icu.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/jdtCompilerAdapter.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/jsoup-1.7.1.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/org.eclipse.core.contenttype.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/org.eclipse.core.jobs.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/org.eclipse.core.resources.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/org.eclipse.core.runtime.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/org.eclipse.equinox.common.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/org.eclipse.equinox.preferences.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/org.eclipse.jdt.core.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/org.eclipse.osgi.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/org.eclipse.text.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/org.netbeans.swing.outline.jar!/" />
    <libelement value="jar://$MODULE_DIR$/mode/ST-4.3.4.jar!/" />
    <libelement value="jar://$MODULE_DIR$/test/lib/junit-4.8.1.jar!/" />
    <src_description expected_position="0">
      <src_folder value="file://$MODULE_DIR$/generated" expected_position="0" />
      <src_folder value="file://$MODULE_DIR$/src" expected_position="1" />
    </src_description>
  </component>
  <component name="NewModuleRootManager">
    <output url="file://$MODULE_DIR$/bin" />
    <exclude-output />
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$/generated" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/src" isTestSource="false" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
    <orderEntry type="module" module-name="processing4-app" />
    <orderEntry type="module" module-name="processing4-core" />
    <orderEntry type="module-library">
      <library name="antlr-4.7.2-complete.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/antlr-4.7.2-complete.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="classpath-explorer-1.0.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/classpath-explorer-1.0.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="com.ibm.icu.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/com.ibm.icu.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="jdtCompilerAdapter.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/jdtCompilerAdapter.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="jsoup-1.7.1.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/jsoup-1.7.1.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.core.contenttype.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.core.contenttype.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.core.jobs.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.core.jobs.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.core.resources.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.core.resources.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.core.runtime.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.core.runtime.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.equinox.common.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.equinox.common.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.equinox.preferences.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.equinox.preferences.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.jdt.core.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.jdt.core.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.osgi.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.osgi.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.text.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.text.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.netbeans.swing.outline.jar">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.netbeans.swing.outline.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.lsp4j">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.lsp4j.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.lsp4j.jar!/" />
        </SOURCES>
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="org.eclipse.lsp4j.jsonrpc">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/org.eclipse.lsp4j.jsonrpc.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
    <orderEntry type="module-library">
      <library name="gson">
        <CLASSES>
          <root url="jar://$MODULE_DIR$/mode/gson.jar!/" />
        </CLASSES>
        <JAVADOC />
        <SOURCES />
      </library>
    </orderEntry>
  </component>
</module>