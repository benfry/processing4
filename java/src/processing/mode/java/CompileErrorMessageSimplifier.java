/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
Part of the Processing project - http://processing.org
Copyright (c) 2012-15 The Processing Foundation

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2
as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software Foundation, Inc.
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/

package processing.mode.java;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.util.TreeMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.eclipse.jdt.core.compiler.IProblem;
import org.eclipse.jdt.internal.compiler.problem.DefaultProblem;

import processing.app.Language;
import processing.app.Messages;
import processing.core.PApplet;
import processing.data.StringList;


public class CompileErrorMessageSimplifier {
  /**
   * Mapping between ProblemID constant and the constant name. Holds about 650
   * of them. Also, this is just temporary, will be used to find the common
   * error types, cos you know, identifying String names is easier than
   * identifying 8 digit int constants!
   * TODO: this is temporary
   */
  private static TreeMap<Integer, String> constantsMap;

  private static final boolean DEBUG = false;

  private static final Pattern tokenRegExp = Pattern.compile("\\b token\\b");

  private static void prepareConstantsList() {
    constantsMap = new TreeMap<>();
    Class<DefaultProblem> probClass = DefaultProblem.class;
    Field[] f = probClass.getFields();
    for (Field field : f) {
      if (Modifier.isStatic(field.getModifiers()))
        try {
          if (DEBUG) {
            Messages.log(field.getName() + " :" + field.get(null));
          }
          Object val = field.get(null);
          if (val instanceof Integer) {
            constantsMap.put((Integer) (val), field.getName());
          }
        } catch (Exception e) {
          e.printStackTrace();
          break;
        }
    }
    if (DEBUG) {
      Messages.log("Total items: " + constantsMap.size());
    }
  }


  public static String getIDName(int id) {
    if (constantsMap == null) {
      prepareConstantsList();
    }
    return constantsMap.get(id);
  }


  /**
   * Tones down the jargon in the ecj reported errors.
   */
  public static String getSimplifiedErrorMessage(IProblem iprob, String badCode) {
    if (iprob == null) return null;

    String[] args = iprob.getArguments();

    if (DEBUG) {
      Messages.log("Simplifying message: " + iprob.getMessage() +
                       " ID: " + getIDName(iprob.getID()));
      Messages.log("Arg count: " + args.length);
      for (String arg : args) {
        Messages.log("Arg " + arg);
      }
      Messages.log("Bad code: " + badCode);
    }

    String result = null;

    switch (iprob.getID()) {

      case IProblem.ParsingError:
        if (args.length > 0) {
          result = Language.interpolate("editor.status.error_on", args[0]);
        }
        break;

      case IProblem.ParsingErrorDeleteToken:
        if (args.length > 0) {
          if (args[0].equalsIgnoreCase("Invalid Character")) {
            result = getErrorMessageForCurlyQuote(badCode);
          }
        }
        break;

      case IProblem.ParsingErrorDeleteTokens:
        result = getErrorMessageForCurlyQuote(badCode);
        if (result == null) {
          result = Language.interpolate("editor.status.error_on", args[0]);
        }
        break;

      case IProblem.ParsingErrorInsertToComplete:
        if (args.length > 0) {
          if (args[0].length() == 1) {
            result = getErrorMessageForBracket(args[0].charAt(0));

          } else {
            if (args[0].equals("AssignmentOperator Expression")) {
              result = Language.interpolate("editor.status.missing.add", "=");

            } else if (args[0].equalsIgnoreCase(") Statement")) {
              result = getErrorMessageForBracket(args[0].charAt(0));

            } else {
              result = Language.interpolate("editor.status.error_on", args[0]);
            }
          }
        }
        break;

      case IProblem.ParsingErrorInvalidToken:
        if (args.length > 0) {
          if (args[0].equals("int")) {
            if (args[1].equals("VariableDeclaratorId")) {
              result = Language.text("editor.status.reserved_words");
            } else {
              result = Language.interpolate("editor.status.error_on", args[0]);
            }
          } else if (args[0].equalsIgnoreCase("Invalid Character")) {
            result = getErrorMessageForCurlyQuote(badCode);
          }
          if (result == null) {
            result = Language.interpolate("editor.status.error_on", args[0]);
          }
        }
        break;

      case IProblem.ParsingErrorInsertTokenAfter:
        if (args.length > 0) {
          if (args[1].length() == 1) {
            result = getErrorMessageForBracket(args[1].charAt(0));
          } else {
            // https://github.com/processing/processing/issues/3104
            if (args[1].equalsIgnoreCase("Statement")) {
              result = Language.interpolate("editor.status.error_on", args[0]);
            } else {
              result =
                  Language.interpolate("editor.status.error_on", args[0]) + " " +
                      Language.interpolate("editor.status.missing.add", args[1]);
            }
          }
        }
        break;

      case IProblem.ParsingErrorReplaceTokens:
        result = getErrorMessageForCurlyQuote(badCode);

      case IProblem.UndefinedConstructor:
        if (args.length == 2) {
          String constructorName = args[0];
          // For messages such as "contructor sketch_name.ClassXYZ() is undefined", change
          // constructor name to "ClassXYZ()". See #3434
          if (constructorName.contains(".")) {
            // arg[0] contains sketch name twice: sketch_150705a.sketch_150705a.Thing
            constructorName = constructorName.substring(constructorName.indexOf('.') + 1);
            constructorName = constructorName.substring(constructorName.indexOf('.') + 1);
          }
          String constructorArgs = removePackagePrefixes(args[args.length - 1]);
          result = Language.interpolate("editor.status.undefined_constructor", constructorName, constructorArgs);
        }
        break;

      case IProblem.UndefinedMethod:
        if (args.length > 2) {
          String methodName = args[args.length - 2];
          String methodArgs = removePackagePrefixes(args[args.length - 1]);
          result = Language.interpolate("editor.status.undefined_method", methodName, methodArgs);
        }
        break;

      case IProblem.ParameterMismatch:
        if (args.length > 3) {
          // 2nd arg is method name, 3rd arg is correct param list
          if (args[2].trim().length() == 0) {
            // the case where no params are needed.
            result = Language.interpolate("editor.status.empty_param", args[1]);

          } else {
            result = Language.interpolate("editor.status.wrong_param",
                                          args[1], args[1], removePackagePrefixes(args[2]));
          }
        }
        break;

      case IProblem.UndefinedField:
        if (args.length > 0) {
          result = Language.interpolate("editor.status.undef_global_var", args[0]);
        }
        break;

      case IProblem.UndefinedType:
        if (args.length > 0) {
          result = Language.interpolate("editor.status.undef_class", args[0]);
        }
        break;

      case IProblem.UnresolvedVariable:
        if (args.length > 0) {
          result = Language.interpolate("editor.status.undef_var", args[0]);
        }
        break;

      case IProblem.UndefinedName:
        if (args.length > 0) {
          final String undefinedName = args[0];
          if (undefinedName.equals("frame")) {
            result = Language.interpolate("editor.status.item_removed", undefinedName, "surface");
          }
          else {
            result = Language.interpolate("editor.status.undef_name", undefinedName);
          }
        }
        break;

      case IProblem.UnterminatedString:
        if (badCode.contains("“") || badCode.contains("”")) {
          result = Language.interpolate("editor.status.unterm_string_curly",
              badCode.replaceAll("[^“”]", ""));
        }
        break;

      case IProblem.TypeMismatch:
        if (args.length > 1) {
          result = Language.interpolate("editor.status.type_mismatch", args[0], args[1]);
        }
        break;

      case IProblem.LocalVariableIsNeverUsed:
        if (args.length > 0) {
          result = Language.interpolate("editor.status.unused_variable", args[0]);
        }
        break;

      case IProblem.UninitializedLocalVariable:
        if (args.length > 0) {
          result = Language.interpolate("editor.status.uninitialized_variable", args[0]);
        }
        break;

      case IProblem.AssignmentHasNoEffect:
        if (args.length > 0) {
          result = Language.interpolate("editor.status.no_effect_assignment", args[0]);
        }
        break;

      case IProblem.HidingEnclosingType:
        if (args.length > 0) {
          result = Language.interpolate("editor.status.hiding_enclosing_type", args[0]);
        }
        break;
    }

    if (result == null) {
      String message = iprob.getMessage();
      if (message != null) {
        // Remove all instances of token
        // "Syntax error on token 'blah', delete this token"
        Matcher matcher = tokenRegExp.matcher(message);
        message = matcher.replaceAll("");
        result = message;
      }
    }

    if (DEBUG) {
      Messages.log("Simplified Error Msg: " + result);
    }

    return result;
  }


  /**
   * Converts java.lang.String into String, etc
   */
  static private String removePackagePrefixes(String input) {
    if (!input.contains(".")) {
      return input;
    }
    String[] names = PApplet.split(input, ',');
    StringList result = new StringList();
    for (String name : names) {
      int dot = name.lastIndexOf('.');
      if (dot >= 0) {
        name = name.substring(dot + 1);
      }
      result.append(name);
    }
    return result.join(", ");
  }


  static private String getErrorMessageForBracket(char c) {
    switch (c) {
      case ';': return Language.text("editor.status.missing.semicolon");
      case '[': return Language.text("editor.status.missing.left_sq_bracket");
      case ']': return Language.text("editor.status.missing.right_sq_bracket");
      case '(': return Language.text("editor.status.missing.left_paren");
      case ')': return Language.text("editor.status.missing.right_paren");
      case '{': return Language.text("editor.status.missing.left_curly_bracket");
      case '}': return Language.text("editor.status.missing.right_curly_bracket");
    }
    // This seems to be unreachable and wasn't in PDE.properties.
    // I've added it for 3.0a8, but that seems gross. [fry]
    return Language.interpolate("editor.status.missing.default", c);
  }


  /**
   * @param badCode The code which may contain curly quotes
   * @return Friendly error message if there is a curly quote in badCode,
   *         null otherwise.
   */
  static private String getErrorMessageForCurlyQuote(String badCode) {
    if (badCode.contains("‘") || badCode.contains("’") ||
        badCode.contains("“") || badCode.contains("”")) {
      return Language.interpolate("editor.status.bad_curly_quote",
          badCode.replaceAll("[^‘’“”]", ""));
    } else return null;
  }


//  static private final String q(Object quotable) {
//    return "\"" + quotable + "\"";
//  }


//  static private final String qs(Object quotable) {
//    return " " + q(quotable);
//  }
}
