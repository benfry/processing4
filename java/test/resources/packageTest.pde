import java.util.*;

List<String> test = new ArrayList<>();
