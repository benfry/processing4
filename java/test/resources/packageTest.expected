package test.subtest;

import processing.core.*;
import processing.data.*;
import processing.event.*;
import processing.opengl.*;

import java.util.*;
import java.util.HashMap;
import java.util.ArrayList;
import java.io.File;
import java.io.BufferedReader;
import java.io.PrintWriter;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.IOException;

public class packageTest extends PApplet {

    public void setup() {
List<String> test = new ArrayList<>();
    noLoop();
    }

    static public void main(String[] passedArgs) {
        String[] appletArgs = new String[] { "packageTest" };
        if (passedArgs != null) {
            PApplet.main(concat(appletArgs, passedArgs));
        } else {
            PApplet.main(appletArgs);
        }
    }
}
